File: src/scripts/salary_slip.js
// Copyright (c) 2019, Libermatic and contributors
// For license information, please see license.txt

export const salary_slip = {
  set_training_data: async function(frm) {
    await frappe.call({
      method:
        'psd_customization.fitness_world.api.salary_slip.set_trainings_in_salary_slip',
      args: { doc_json: frm.doc, set_in_response: 1 },
    });
  },
  calculate_training_earnings: async function(frm) {
    const {
      salary_structure,
      total_training_months = 0,
      training_rate = 0,
    } = frm.doc;
    const {
      message: { training_salary_component } = {},
    } = await frappe.db.get_value(
      'Salary Structure',
      salary_structure,
      'training_salary_component'
    );
    const earning = frm.doc.earnings.find(
      ({ salary_component }) => salary_component === training_salary_component
    );
    if (earning) {
      earning.amount = total_training_months * training_rate;
      frm.refresh_field('earnings');
    }
  },
  toggle_training_section: function(frm) {
    frm.toggle_display(
      ['training_section', 'trainings'],
      cint(frm.doc.salary_slip_based_on_training) === 1
    );
  },
};

function calculate_training_months(frm) {
  frm.set_value(
    'actual_training_months',
    frm.doc.trainings
      .map(({ months = 0.0 }) => months)
      .reduce((a, x = 0) => a + x, 0)
  );
  frm.set_value(
    'total_training_months',
    frm.doc.trainings
      .map(
        ({ months = 0.0, cost_multiplier = 1.0 }) => months * cost_multiplier
      )
      .reduce((a, x = 0) => a + x, 0)
  );
}

export const salary_slip_training = {
  training: async function(frm, cdt, cdn) {
    const { training } = frappe.get_doc(cdt, cdn);
    if (training) {
      const no_of_occurences = frm.doc.trainings.filter(
        ({ training: t }) => t === training
      ).length;
      if (no_of_occurences > 1) {
        frm.get_field('trainings').grid.grid_rows_by_docname[cdn].remove();
      } else {
        const { start_date, end_date } = frm.doc;
        const {
          message: { months, gym_subscription, cost_multiplier } = {},
        } = await frappe.call({
          method:
            'psd_customization.fitness_world.api.salary_slip.get_training_data',
          args: { training, start_date, end_date },
        });
        frappe.model.set_value(cdt, cdn, 'months', months);
        frappe.model.set_value(cdt, cdn, 'subscription', gym_subscription);
        frappe.model.set_value(cdt, cdn, 'cost_multiplier', cost_multiplier);
      }
    } else {
      ['months', 'subscription'].forEach(field => {
        frappe.model.set_value(cdt, cdn, field, null);
      });
    }
  },
  months: function(frm, cdt, cdn) {
    calculate_training_months(frm);
  },
  trainings_remove: function(frm, cdt, cdn) {
    calculate_training_months(frm);
  },
};

export default { salary_slip, salary_slip_training };

File: psd_customization/fitness_world/api/salary_slip.py
# -*- coding: utf-8 -*-
# Copyright (c) 2019, Libermatic and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.query_builder import Interval
from frappe.query_builder.functions import IfNull
from frappe.utils import getdate, add_days, cint, flt
from builtins import str
import json
from functools import partial
from toolz import merge, pluck, compose

from psd_customization.utils.fp import mapr


@frappe.whitelist()
def set_trainings_in_salary_slip(doc_json, set_in_response=0):
    doc = (
        frappe.get_doc(json.loads(doc_json)) if isinstance(doc_json, str) else doc_json
    )
    if not doc.salary_structure:
        joining_date, relieving_date = frappe.db.get_value(
            "Employee", doc.employee, ["date_of_joining", "relieving_date"]
        )
        doc.salary_structure = doc.check_sal_struct(joining_date, relieving_date)
    structure = frappe.get_doc("Salary Structure", doc.salary_structure)
    doc.salary_slip_based_on_training = structure.salary_slip_based_on_training
    doc.set("trainings", [])
    if doc.salary_slip_based_on_training:
        trainings = get_trainings_for_salary_slip(doc.employee, doc.end_date)
        for row in trainings:
            doc.append(
                "trainings",
                {
                    "training": row.name,
                    "member_name": row.gym_member_name,
                    "months": row.months,
                    "subscription": row.gym_subscription,
                    "cost_multiplier": row.cost_multiplier or 1.0,
                },
            )
        doc.actual_training_months = compose(sum, partial(pluck, "months"))(trainings)
        doc.total_training_months = compose(
            sum,
            partial(map, lambda t: flt(t.months or 0) * flt(t.cost_multiplier or 1)),
        )(trainings)
        doc.training_rate = structure.training_monthly_rate
        add_earning_for_training(
            doc,
            structure.training_salary_component,
            doc.total_training_months * doc.training_rate,
        )
    if cint(set_in_response):
        frappe.response.docs.append(doc)


def get_trainings_for_salary_slip(employee, end_date):
    trainer = frappe.db.exists("Gym Trainer", {"employee": employee})
    if not trainer:
        return []

    TrainerAllocation = frappe.qb.DocType("Trainer Allocation")
    GymSubscription = frappe.qb.DocType("Gym Subscription")
    trainings = (
        frappe.qb.from_(TrainerAllocation)
        .left_join(GymSubscription)
        .on(GymSubscription.name == TrainerAllocation.gym_subscription)
        .select(
            TrainerAllocation.name,
            GymSubscription.member_name.as_("gym_member_name"),
            GymSubscription.name.as_("gym_subscription"),
            TrainerAllocation.salary_till,
            TrainerAllocation.from_date,
            TrainerAllocation.to_date,
            GymSubscription.day_fraction,
            GymSubscription.cost_multiplier,
        )
        .where(
            (TrainerAllocation.gym_trainer == trainer)
            & (
                IfNull(
                    TrainerAllocation.salary_till,
                    TrainerAllocation.from_date - Interval(days=1),
                )
                < end_date
            )
            & (
                TrainerAllocation.to_date
                - IfNull(
                    TrainerAllocation.salary_till,
                    TrainerAllocation.from_date - Interval(days=1),
                )
                > 0
            )
        )
    ).run(as_dict=1)
    return mapr(_set_days(end_date), trainings)


def add_earning_for_training(doc, salary_component, amount):
    for row in doc.earnings or []:
        if row.salary_component == salary_component:
            row.amount = amount
            return
    doc.append(
        "earnings",
        {
            "salary_component": salary_component,
            "abbr": frappe.db.get_value(
                "Salary Component", salary_component, "salary_component_abbr"
            ),
            "amount": amount,
        },
    )


def _set_days(end_date):
    def fn(row):
        from_date = add_days(row.salary_till, 1) if row.salary_till else row.from_date
        to_date = min(getdate(end_date), row.to_date)
        days = (to_date - from_date).days + 1
        return frappe._dict(merge(row, {"months": days * row.day_fraction}))

    return fn


@frappe.whitelist()
def get_training_data(training, start_date, end_date):
    trainer_alloc = frappe.get_doc("Trainer Allocation", training)
    if trainer_alloc:
        ta = _set_days(end_date)(frappe._dict(trainer_alloc.as_dict()))
        cost_multiplier = frappe.db.get_value(
            "Gym Subscription", ta.gym_subscription, "cost_multiplier"
        )
        return {
            "months": ta.months,
            "gym_subscription": ta.gym_subscription,
            "cost_multiplier": cost_multiplier or 1.0,
        }


def training_query(doctype, txt, searchfield, start, page_len, filters):
    trainer = frappe.db.exists("Gym Trainer", {"employee": filters.get("employee")})
    if not trainer:
        return []

    TrainerAllocation = frappe.qb.DocType("Trainer Allocation")
    return (
        frappe.qb.from_(TrainerAllocation)
        .select(TrainerAllocation.name, TrainerAllocation.gym_member_name)
        .where(
            (TrainerAllocation.gym_trainer == trainer)
            & (
                TrainerAllocation.salary_till.isnull()
                | (TrainerAllocation.salary_till < TrainerAllocation.to_date)
            )
        )
        .orderby(TrainerAllocation.to_date)
        .limit(page_len)
        .offset(start)
    ).run()
